Plutarch/Api/V1/AssocMap.hs

```diff
@@ -56,3 +63,37 @@ instance
       x' <- Plutus.fromData x
       y' <- Plutus.fromData y
       Just (x', y')
+
+-- | Look up the given key in a 'PMap'.
+lookup :: (PIsData k, PIsData v) => Term (s :: S) (k :--> PMap k v :--> PMaybe v)
+lookup = phoistAcyclic $
+  plam $ \key ->
+    lookupDataWith
+      # (phoistAcyclic $ plam $ \pair -> pcon $ PJust $ pfromData $ psndBuiltin # pair)
+      # pdata key
+
+-- | Look up the given key data in a 'PMap'.
+lookupData :: (PIsData k, PIsData v) => Term (s :: S) (PAsData k :--> PMap k v :--> PMaybe (PAsData v))
+lookupData = lookupDataWith # (phoistAcyclic $ plam $ \pair -> pcon $ PJust $ psndBuiltin # pair)
+
+-- | Look up the given key data in a 'PMap', applying the given function to the found key-value pair.
+lookupDataWith ::
+  (PIsData k, PIsData v) =>
+  Term
+    (s :: S)
+    ( (PBuiltinPair (PAsData k) (PAsData v) :--> PMaybe x)
+        :--> PAsData k
+        :--> PMap k v
+        :--> PMaybe x
+    )
+lookupDataWith = phoistAcyclic $
+  plam $ \unwrap key map ->
+    pmaybe # pcon PNothing # unwrap
+      #$ plet (plam $ \pair -> pfstBuiltin # pair #== key)
+      $ \predicate ->
+        pfind # predicate # pto map
+
+-- | Construct a singleton 'PMap' with the given key and value.
+singleton :: (PIsData k, PIsData v) => Term (s :: S) (k :--> v :--> PMap k v)
```

Comment:
potentially better to take in `PAsData k` and `PAsData v` to avoid the extraneous `pdata` calls within.

Reply:
Good point, but I've made that a separate function `singletonData`.